src/sentry/incidents/models/alert_rule.py

```diff
@@ -564,7 +578,7 @@ class AlertRuleTriggerAction(AbstractNotificationAction):
     Type = ActionService
     TargetType = ActionTarget
 
-    _factory_registrations: dict[ActionService, ActionHandlerFactory] = {}
+    factory_registrations = FactoryRegistry()
```

Comment:
Previously, the privacy-by-convention of the leading underscore was being honored only in prod code but ignored by several unit tests. Now those unit tests interact with it only through the new `TemporaryAlertRuleTriggerActionRegistry` class from testutils.

I'm on the fence about whether to keep this as underscore-prefixed, and to just let `TemporaryAlertRuleTriggerActionRegistry` be an exception (since it's only in one place and relatively easy to fix if we break it again).

Reply:
Are we using this class member variable as a singleton? What's the benefit of this over a module-level singleton we reference instead?

Regardless, I think keeping the `_` prefix here makes sense just to make it clear that mutating this is discouraged.

Reply:
> Are we using this class member variable as a singleton?

Yes. We previously had a single global `dict[ActionService, ActionHandlerFactory]` object, and now a `FactoryRegistry` instance is replacing it (and the only reason I made `FactoryRegistry` its own class is because I wanted to have a single `register` method responsible for populating two dicts).

> Regardless, I think keeping the `_` prefix here makes sense just to make it clear that mutating this is discouraged.

I agree. I think I'll put the underscore back and add a comment that `TemporaryAlertRuleTriggerActionRegistry` is a testutil with special dispensation to violate privacy.

> What's the benefit of this over a module-level singleton we reference instead?

The distinction matters less without the underscore prefix. If it's private to the class, that communicates that it should be modified only by calling `classmethod`s on `AlertRuleTriggerAction`.

However (is this what you were getting at?) `FactoryRegistry` would make sense as a *public* module-level singleton. I think the real problem is that (prior to any of my factoring) the `AlertRuleTriggerAction` class was pulling double duty as a registry object. Methods like `register_type` want to be instance methods on a registry object, not class methods on anything.

Hence, a good next refactoring step would be to extract *all* the class methods from `AlertRuleTriggerAction` into a new public class, which means going around to all the calls like `@AlertRuleTriggerAction.register_type(...)` and change them to something like `@action_handler_factory_registry.register_type(...)` where `action_handler_factory_registry` is a global, static singleton. (That still leaves us with the pre-existing problems of having global, static singletons in the first place... which, lest we forget, is why I wanted to nuke `STRING_TO_ACTION_TYPE` in the first place. 😜)

Reply:
> However (is this what you were getting at?) FactoryRegistry would make sense as a public module-level singleton. I think the real problem is that (prior to any of my factoring) the AlertRuleTriggerAction class was pulling double duty as a registry object.

Yep, exactly what I was getting at. I see your point about this being very similar to the dictionary you were refactoring away in the first place, but I don't see a great way around having _some_ sort of registration singleton at the moment, unless we pursue a more dynamic way of registering `ActionHandlerFactory` classes similar to what we did with RPC services 🤔

Reply:
https://github.com/getsentry/sentry/pull/74936/commits/44aae2b8bb28bd7a9f52a7fae13661965d79d6a3